Add database branch

LeonovichMariia opened the pull request.

LeonovichMariia committed:
feat: Добавлены новые свойства модели Film, а также новые классы Genre(жанр) и Mpa(рейтинг). Добавлен класс Friend и новое свойство - статус для связи «дружба» между двумя пользователями.

LeonovichMariia committed:
feat: В файл pom.xml добавлены необходимые зависимости, а в application.properties настройки для работы с базой H2

LeonovichMariia committed:
feat: Добавлены класс FilmDbStorage, файлы schema.sql и data.sql. fix: внесены изменения в классы FilmService, InMemoryFilmStorage и InMemoryUserStorage

LeonovichMariia committed:
fix: внесены изменения в ER-диаграмму БД и файл README. feat: Добавлены классы FilmMapper (реализован метод mapRow) и GenresMapper (реализован метод mapRow). В классе FilmDbStorage реализован метод findObjectById.

LeonovichMariia committed:
fix: В методе findObjectById класса FilmDbStorage добавлена проверка на null

LeonovichMariia committed:
feat: Добавлены классы-контроллеры, классы-сервисы, а также классы и интерфейсы для жанров и рейтингов. Реализована логика методов по получению списка всех жанров и рейтингов, а также по идентификатору. Добавлены классы UserMapper с логикой и USerDbStorage, реализованы методы findObjectById и getAllObjects. fix: Внесены небольшие правки в README и ER-диаграмму

LeonovichMariia committed:
feat: В классах GenreDbStorage и MpaDbStorage в методах  get*ById добавлена проверка на null. В класс Film добавлен метод addGenre. В FilmDbStorage реализованы методы addObject, renewalObject, deleteObject, getAllObjects, saveGenresToFilm

LeonovichMariia committed:
fix: Внесены коррективы в ER - диаграмму, описание БД. feat: Объекту Film добавлено новое свойство - rate. Добавлены классы и интерфейсы- хранилища для FilmAudience, FilmGenre, Friend, а также реализована логика этих классов

LeonovichMariia committed:
code style

LeonovichMariia committed:
feat: Добавлена проверка на null в методе findObjectById в классе UserDbStorage, FilmDbStorage GenreDbStorage и MpaDbStorage. В классе FilmDbStorage добавлен batchUpdate в saveGenresToFilm, исправлен sql запрос в getAllObjects. Добавлены тесты : GenreControllerTest и MpaControllerTest.

LeonovichMariia committed:
feat: Добавлены тесты в класс UserControllerTest. fix: code style

LeonovichMariia committed:
delete: Удалены не использующиеся методы из класса Film и не использующиеся классы(FilmGenreStorage и FilmGenreDbStorage).
feat: Добавлены классы InMemoryFilmAudienceStorage и и InMemoryFriendStorage,а также соответствующая логика для сохранения возможности работы в памяти.
fix: внесены изменения в метод  findObjectById класса FilmDbStorage. Исправлен метод getPopularFilmsList класса FilmAudienceDbStorage.

LeonovichMariia committed:
feat: добавлены тесты в FilmControllerTest и UserControllerTest.

LeonovichMariia committed:
feat: добавлены тесты в FilmControllerTest и UserControllerTest.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Friend.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class Friend {
```
Создавать отдельный элемент модели данных для дружбы не требуется, вся дружба должна определяться на основании компонента `User` и таблицы дружбы (таблицы, которая связывает пользователей между собой по факту дружбы, как односторонней, так и обоюдной).

Если вдруг не получится сделать обоюдную дружбу без данного компонента, то тогда предлагаю вернуться к реализации дружбы, которая была в прошлом (10) спринте, так как тесты POSTMAN такой вариант допускают.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmAudienceDbStorage.java:
```diff
@@ -0,0 +1,45 @@
+    public void removeLike(long filmId, long userId) {
+        String sql = "DELETE FROM film_audience WHERE film_id = ? AND user_id = ?";
+        jdbcTemplate.update(sql, filmId, userId);
+        String sqlRemoveLike = "UPDATE film_data SET rate = rate - 1 WHERE film_id = ?";
```
Так как `rate` теперь хранится на стороне БД, то можно убрать его из модели данных, там его логика теряет смысл

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,149 @@
+
+        String sqlGetGenresForFilm = "SELECT * FROM film_genre INNER JOIN genre ON film_genre.genre_id=genre.genre_id WHERE film_id=?";
+
+        for (Film film : films) {
```
Если фильмов для обновления несколько (при запросе через getAll), то тогда для каждого фильма будет запрошен список жанров, что порождает в сумме N запросов к БД, надо уместить всё в 1 запрос, для этого предлагаю следующее:

1. Делаете запрос к двум таблицам GENRES, FILM_GENRES - можете сделать как Вам удобно, либо через подзапрос, либо выборку из двух таблиц, либо воспользоваться оператором join, выбирая все требуемые поля
2. Дальше можно воспользоваться специальной версией jdbcTemplate.query() для работы с несколькими объектами (там несколько возможных вариантов, подберите тот который Вам подходит)
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html
Такой вид метода позволит обработать все в одном запросе. Вы передаете туда первым параметром Ваш sql запрос (пункт 1), затем вторым параметром передаёте лямбду, в которой добавляете жанр в фильм через resultSet, а третьим параметром передаете набор идентификаторов фильмов, которые участвуют в запросе.
Таким образом будет выполнен сразу набор операций.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,149 @@
+            return;
+        }
+        List<Genre> genreList = new ArrayList<>(genres);
+        jdbcTemplate.batchUpdate(
```
Очень хорошо, что получилось разобраться с пакетным обновлением!

KirPyt commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,59 @@
+DROP ALL OBJECTS;
```
Желательно прописывать таблицы явно + писать IF EXISTS

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -28,33 +34,50 @@ public void validate(User user) {
+        User friend = getFriend(userId, otherUserId);
+        if (friend != null) {
+            log.info(LogMessages.FRIEND_UNCONFIRMED.toString());
+            Friend unconfirmedFriend = user.getFriends().stream()
```
Стоит отказаться от компонента Friends, как описал в верхних замечаниях и попробовать реализовать всё посредством `User`.

KirPyt commented on src/main/resources/application.properties:
```diff
@@ -1 +1,10 @@
+spring.datasource.password=password
+spring.h2.console.enabled=true
 
+app.storage.type=db
```
Отличное решение вынести тип компонента в файл properties!

LeonovichMariia committed:
delete: удален класс Friend. fix: внесены изменения в метод addFriend с учетом удаления.

LeonovichMariia committed:
feat: В класс GenreDbStorage добавлен метод getGenresForFilms.
fix: code style. В классе FilmDbStorage внесены изменения в метод getAllObjects и и реализован метод mapRow.

LeonovichMariia committed:
code style

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,56 @@
+                    while (rs.next()) {
+                        long filmId = rs.getLong("film_id");
+                        List<Genre> genreList = new ArrayList<>();
+                        genreList.add(new Genre(rs.getLong("genre_id"), rs.getString("genre_name")));
```
Есть мысль, что логика наполнения жанров в данном месте неверная, так как в данный список попадают все жанры, в свою очередь у каждого фильма свой набор жанров. Иными словами из результата запроса к БД сначала отбираются все фильмы с конкретным id, потом отобранные жанры для данного id фильма записываются в список, затем уже кладутся в мапу фильм -> список фильмов.

LeonovichMariia committed:
fix: Внесены изменения в метод getGenresForFilms в классе GenreDbStorage.
feat: добавлен метод mapRow.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/genre/GenreDbStorage.java:
```diff
@@ -0,0 +1,65 @@
+                    while (rs.next()) {
+                        long filmId = rs.getLong("film_id");
+                        List<Genre> genres = filmList.getOrDefault(filmId, new ArrayList<>());
+                        genres.add(mapRow(rs));
```
Данный вариант не учитывает случай, когда у фильма может быть больше 1 жанра.
Надо изменить данный цикл, чтобы сначала список `genres` наполнялся жанрами фильма, а потом уже наполненный список сохранялся в хэш-мапу `filmList`

LeonovichMariia committed:
fix: внсены правки в метод getGenresForFilms в классе GenreDbStorage. + codestyle
